dozer-sql/src/pipeline/product/set/record_map/bloom.rs

```diff
@@ -0,0 +1,195 @@
+//! Based on https://www.arunma.com/2023/03/19/build-your-own-counting-bloom-filter-in-rust/
+
+use std::hash::Hash;
+
+use dozer_types::serde::{Deserialize, Serialize};
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(crate = "dozer_types::serde")]
+pub struct CountingBloomFilter {
+    counters: Vec<u8>,
+    num_hashes: u32,
+    #[serde(skip)]
+    hasher: hash::BloomHasher,
+}
+
+impl CountingBloomFilter {
+    pub fn with_rate(false_positive_rate: f32, expected_num_items: u32) -> Self {
+        let num_counters = optimal_num_counters(expected_num_items, false_positive_rate);
+        let num_hashes = optimal_num_hashes(expected_num_items, num_counters);
+        Self {
+            counters: vec![0; num_counters],
+            num_hashes,
+            hasher: Default::default(),
+        }
+    }
+
+    pub fn insert<V: Hash>(&mut self, value: &V) {
+        for slot in calculate_slots(&self.hasher, value, self.num_hashes, self.counters.len()) {
+            self.counters[slot] = self.counters[slot].saturating_add(1);
+        }
+    }
+
+    pub fn remove<V: Hash>(&mut self, value: &V) {
+        let slots = calculate_slots(&self.hasher, value, self.num_hashes, self.counters.len())
+            .collect::<Vec<_>>();
+        if slots.iter().all(|slot| self.counters[*slot] > 0) {
+            for slot in slots {
+                self.counters[slot] = self.counters[slot].saturating_sub(1);
+            }
+        }
+    }
+
+    pub fn estimate_count<V: Hash>(&self, value: &V) -> u8 {
+        calculate_slots(&self.hasher, value, self.num_hashes, self.counters.len())
+            .map(|slot| self.counters[slot])
+            .min()
+            .unwrap_or(0)
+    }
+
+    pub fn clear(&mut self) {
+        self.counters.iter_mut().for_each(|counter| *counter = 0);
+    }
+}
+
+fn optimal_num_counters(num_items: u32, false_positive_rate: f32) -> usize {
+    -(num_items as f32 * false_positive_rate.ln() / (2.0f32.ln().powi(2))).ceil() as usize
```

Comment:
You should use `f64`s instead of `f32`s to do the computation in this function. It produces different results using the default parameters `10000000` and `0.01` for `num_items` and `false_positive_rate` accordingly; `95850592` for `f32` and `95850584` for `f64`. The reference you linked uses `f64` in this function, as well as the external `bloom` crate in a function that performs a similar calculation.

Reply:
Good catch!